src/main/java/org/springframework/integration/kafka/inbound/KafkaMessageDrivenChannelAdapter.java

```diff
@@ -295,6 +321,21 @@ public int afterShutdown() {
 		return getPhase();
 	}
 
+	@Override
+	public void onApplicationEvent(ListenerContainerIdleEvent event) {
+		Consumer<?, ?> consumer = event.getConsumer();
+		pauseIfNecessary(consumer);
+		if (this.resumePending.getAndSet(false)) {
+			consumer.resume(consumer.paused());
```

Comment:
You know there is no guarantee that the even arrives in the same thread it has been published.
If that is so important to retain calls order in the same thread though...

Reply:
Well, this won't work if they use async publishing, the consumer is not thread-safe.

But you do raise an interesting problem - there might be multiple container threads.

Need to think...

Reply:
Maybe we definitely need to delegate that to the `container` and deal with `pause/resume` state in each `ListenerConsumer` 's main loop ?

Reply:
Yeah - I came to same conclusion (independently 😄 )